subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/annotations/NestedBeanAnnotationHandler.java

```diff
@@ -0,0 +1,169 @@
+import org.gradle.internal.UncheckedException;
+
+import javax.annotation.Nullable;
+import java.lang.annotation.Annotation;
+import java.lang.reflect.InvocationHandler;
+import java.lang.reflect.Proxy;
+
+import static org.gradle.api.internal.tasks.TaskValidationContext.Severity.ERROR;
+
+public class NestedBeanAnnotationHandler implements PropertyAnnotationHandler {
+
+    private NestedBeanResolver<PropertyNode> nestedBeanResolver = new NestedBeanResolver<PropertyNode>();
+
+    @Override
+    public Class<? extends Annotation> getAnnotationType() {
+        return Nested.class;
+    }
+
+    @Override
+    public void visitPropertyValue(PropertyValue propertyValue, PropertyVisitor visitor, PropertySpecFactory specFactory, NestedBeanContext<PropertyNode> context) {
+        Object nested;
+        try {
+            nested = propertyValue.getValue();
+        } catch (Exception e) {
+            visitor.visitInputProperty(specFactory.createInputPropertySpec(propertyValue.getPropertyName(), new InvalidPropertyValue(e)));
+            return;
+        }
+        if (nested != null) {
+            nestedBeanResolver.resolve(
+                context.createNode(propertyValue.getPropertyName(), nested),
+                new ImplementationAddingPropertyContext(context, visitor, specFactory));
+        } else if (!propertyValue.isOptional()) {
+            visitor.visitInputProperty(specFactory.createInputPropertySpec(propertyValue.getPropertyName(), new AbsentPropertyValue()));
+        }
+    }
+
+    private static class ImplementationAddingPropertyContext implements NestedBeanContext<PropertyNode> {
+        private final NestedBeanContext<PropertyNode> delegate;
+        private final PropertyVisitor visitor;
+        private final PropertySpecFactory specFactory;
+
+        public ImplementationAddingPropertyContext(NestedBeanContext<PropertyNode> delegate, PropertyVisitor visitor, PropertySpecFactory specFactory) {
+            this.delegate = delegate;
+            this.visitor = visitor;
+            this.specFactory = specFactory;
+        }
+
+        @Override
+        public PropertyNode createNode(String propertyName, Object nested) {
+            return delegate.createNode(propertyName, nested);
+        }
+
+        @Override
+        public void addNested(PropertyNode node) {
+            visitImplementation(node, visitor, specFactory);
+            delegate.addNested(node);
+        }
+
+        @Override
+        public boolean isIterable(PropertyNode node) {
+            return delegate.isIterable(node);
+        }
+    }
+
+    private static void visitImplementation(PropertyNode node, PropertyVisitor visitor, PropertySpecFactory specFactory) {
+        // The root bean (Task) implementation is currently tracked separately
+        DefaultTaskInputPropertySpec implementation = specFactory.createInputPropertySpec(node.getQualifiedPropertyName("class"), new ImplementationPropertyValue(getImplementationClass(node.getBean())));
```

Comment:
It is okay to evaluate the bean class here as opposed to only calculating it lazily in `ImplementationPropertyValue.call()`, because we'd recalculate the properties anyway should the type change, right?

Reply:
Yeah, the nested bean itself is fixed for one property calculation and we do not detect changes in that bean.